futures/src/supervisor.rs

```diff
@@ -0,0 +1,843 @@
+        } else if self.caught_unexpected_exit {
+            Err(Error::UnexpectedExit)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn handle_message(&mut self, message: &SupervisorMessage) -> ControlFlow<(), ()> {
+        match message {
+            SupervisorMessage::TaskAdded => {
+                self.active_tasks += 1;
+                ControlFlow::Continue(())
+            }
+            SupervisorMessage::TaskFinished { panic } => {
+                self.active_tasks -= 1;
+                if *panic {
+                    self.caught_panic = true;
+                    if !self.shutdown_signal.is_sent() {
+                        iroha_logger::error!("Some task panicked, shutting down everything...");
+                        self.shutdown_signal.send();
+                    }
+                } else if !self.shutdown_signal.is_sent() {
+                    self.caught_unexpected_exit = true;
+                    iroha_logger::error!(
+                        "Some task exited unexpectedly, shutting down everything..."
+                    );
+                    self.shutdown_signal.send();
+                }
+                if self.active_tasks == 0 {
+                    ControlFlow::Break(())
+                } else {
+                    ControlFlow::Continue(())
+                }
+            }
+            SupervisorMessage::FinishIfEmpty => {
+                if self.active_tasks == 0 {
+                    ControlFlow::Break(())
+                } else {
+                    ControlFlow::Continue(())
+                }
+            }
+        }
+    }
+}
+
+impl Default for Supervisor {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// TODO: shall we
+
+// impl Drop for Supervisor {
+//     fn drop(&mut self) {
+//         // TODO: fire shutdown signal?
+//         self.task_handle.abort();
+//     }
+// }
+
+impl Supervisor {
+    /// Create a new supervisor.
+    ///
+    /// This must be executed within a tokio runtime.
+    pub fn new() -> Self {
+        let (task_tx, task_rx) = mpsc::channel(u8::MAX as usize);
+        let shutdown_signal = ShutdownSignal::new();
+
+        let task_handle = tokio::spawn(SupervisorTask::new(task_rx, shutdown_signal.clone()).run());
+
+        Self {
+            task_handle,
+            task_tx,
+            shutdown_signal,
+            monitoring_some: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    /// Get a copy of the supervisor's shutdown signal
+    pub fn shutdown_signal(&self) -> ShutdownSignal {
+        self.shutdown_signal.clone()
+    }
+
+    /// Monitors a given [`Child`].
+    ///
+    /// When it panics, it triggers [`Self::shutdown`] and waits until handles
+    /// marked with [`OnShutdown::Wait`] complete within a given timeout. When all such handles
+    /// complete (or timeout elapsed), it aborts all incomplete handles.
+    #[track_caller]
+    pub fn monitor(&self, child: impl Into<Child>) {
```

Comment:
I did it to avoid `async`-coloring of all of the code that uses `monitor`. Though, it was on intermediate implementation stage, I may reconsider it now.

Reply:
Not relevant anymore: now `Supervisor` doesn't spawn anything until `start()`, thus doesn't need to send anything in `monitor()`